app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt

```diff
@@ -19,36 +24,138 @@ interface PagedHistoryProvider {
      * @param numberOfItems How many items to fetch
      * @param onComplete A callback that returns the list of [VisitInfo]
      */
-    fun getHistory(offset: Long, numberOfItems: Long, onComplete: (List<VisitInfo>) -> Unit)
+    fun getHistory(offset: Long, numberOfItems: Long, onComplete: (List<HistoryItem>) -> Unit)
 }
 
-// A PagedList DataSource runs on a background thread automatically.
-// If we run this in our own coroutineScope it breaks the PagedList
-fun HistoryStorage.createSynchronousPagedHistoryProvider(): PagedHistoryProvider {
-    return object : PagedHistoryProvider {
-
-        override fun getHistory(
-            offset: Long,
-            numberOfItems: Long,
-            onComplete: (List<VisitInfo>) -> Unit
-        ) {
-            runBlockingIncrement {
-                val history = getVisitsPaginated(
-                    offset,
-                    numberOfItems,
-                    excludeTypes = listOf(
-                        VisitType.NOT_A_VISIT,
-                        VisitType.DOWNLOAD,
-                        VisitType.REDIRECT_TEMPORARY,
-                        VisitType.RELOAD,
-                        VisitType.EMBED,
-                        VisitType.FRAMED_LINK,
-                        VisitType.REDIRECT_PERMANENT
+/**
+ * @param historyStorage
+ */
+class DefaultPagedHistoryProvider(
+    private val historyStorage: PlacesHistoryStorage
+) : PagedHistoryProvider {
+
+    @Suppress("LongMethod")
+    override fun getHistory(
+        offset: Long,
+        numberOfItems: Long,
+        onComplete: (List<HistoryItem>) -> Unit
+    ) {
+        // A PagedList DataSource runs on a background thread automatically.
+        // If we run this in our own coroutineScope it breaks the PagedList
+        runBlockingIncrement {
+            if (FeatureFlags.showHistorySearchGroups) {
+                val historyMetadata: MutableList<HistoryMetadata> =
+                    historyStorage.getHistoryMetadataSince(Long.MIN_VALUE)
+                        .filter { it.totalViewTime > 0 && it.key.searchTerm != null }
+                        .toMutableList()
+                val historyGroups: MutableMap<String, List<HistoryMetadata>> = historyMetadata
```

Comment:
I think this is a decent first pass 👍 

Let's assume for now that we only care about combing regular history with search-term groups.

First, let's outline what you're currently doing (just for my own understanding), for each "data page" load:
- load all metadata that we have in the db that have a searchTerm
- compute search groups - metadata.groupBy(searchTerm).
- load a page of regular history
- results = []
- for each item in fetched history:
  - if there is no matching-by-url metadata, results.add(item); otherwise:
  - find a search group this item belongs to,
  - results.add(group)
  - remove the search group
  - remove the matched-by-url metadata item

Functionally, there's one main problem with this: you're not taking into account `visitedAt` timestamps for the group. In your logic, if you had a group from two weeks ago, and then also had a visit from today to a url that's present in the group, that group will show-up on the first page. I think that's a bug. 

This logic also doesn't help us with deletes - e.g. how will we go in the future from a metadata item to a visit? For each metadata delete, we'll need to perform two calls - deleteMetadata(key), deleteVisit(url, timestamp).

This is also doing some unnecessary work - mostly that you're fetching metadata every page load; iterating through visits and performing metadata lookups for each item is more work than necessary (e.g. visits >= metadata in groups).
 
Besides these things, it feels like it should generally do the right thing and is certainly a good first pass 👍

Personally, I'd reverse this logic (iterate over groups) and take into account visitedAt timestamps. It will also be great to maintain a Metadata -> Visit mapping to allow deletions, e.g. at the class-level, assuming we can somehow access this mapping when we handle deletions in HistoryFragment..?

- At the class-level of `DefaultPagedHistoryProvider`, fetch all searchTerm groups and transform them into `HistoryItem` objects. When creating a group, make sure that its `visitedAt` is the timestamp of the latest visit in it. You only need to do this fetch+transform operation once, I think.
- for each page of history items, find the visitedAt range - lowest/highest timestamps of visits you got back
- groupsInRange = allGroups.filter(visitedAt within range)
- for group in groupsInRange:
  - for metadata in group:
    - visits.filter(url), find closest matching visit by timestamp (here, we're assuming that metadata are created very near-in-time to visits, which I think is true)
    - metadataToVisitMap[metadata] = matchedVisit
    - remove matchedVisit from the visits list
- results.addAll(groupsInRange)
- results.addAll(visits) (just the visits w/o corresponding metadata at this point)
- results.sortBy(item.visitedAt)

This has at least one edge-case I can think of: a group may have items that won't appear on the current results page. E.g. if most items in the group were visited a few days ago, but the most recent item in the group was visited today, then we won't fully populated metadataToVisitMap, and when we get to the history page with these items, we will display them on the screen (even though they were in an earlier group already).

I think that it's fine to have this edge case for now. It's almost a feature, since we're still displaying old items in groups chronologically where they belong. It's not great for deletions though, since user may need to remove a visit twice...

You can fix this edge case by maintaining a list of "left-over" metadata items that didn't match, and trying to match them up in subsequent passes as you go.

Another edge case: the final `results` may have less items than what was requested via `numberOfItems`. I think that's okay to start with..? We could also just do multiple passes - e.g. if at the end we don't have "enough" results, fetch another history page.
From the [android docs](https://developer.android.com/reference/android/arch/paging/ItemKeyedDataSource.LoadParams), `Returned page can be of this size, but it may be altered if that is easier, e.g. a network data source where the backend defines page size. `, so it seems like it's fine to return less or more than requested.